Step2 - 경로 조회 기능

ssamzag opened the pull request:
안녕하세요 오태경 리뷰어님!
Step2 - 경로 조회 기능 리뷰 요청드려요.

이번 리뷰도 잘 부탁드립니다! 😄

ssamzag committed:
feat: 클래스 파일 추가

ssamzag committed:
feat&refactor: LineService 에 StationRepository 등록 관련 수정

ssamzag committed:
feat&refactor: 지하철 경로 조회 기능 추가 외
- Sections.java(refactor): 생성자 및 add 메서드 방어적 복사 리팩터링
- feat: 지하철 경로 조회 기능 api 및 도메인 추가
- test: 지하철 경로 기능 관련 도메인 및 인수 테스트

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Section.java:
```diff
@@ -80,4 +80,9 @@ public boolean existUpStation() {
         return this.distance.plus(section.distance);
     }
+
+    public boolean containStations(Station station, Station nextStation) {
+        return equalsUpStation(station) && equalsDownStation(nextStation)
+                || equalsUpStation(nextStation) && equalsDownStation(station);
```
여러 조건을 하나의 메서드로 추출하여 의미 전달 👍

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -27,6 +32,17 @@ public void add(Line line, Station upStation, Station downStation, Distance dist
 
+    public void add(List<Section> sections) {
+        this.sections.addAll(new ArrayList<>(sections));
+    }
```
해당 메서드는 사용하지 않는 것 같아요 :)

ssamzag replied:
여러방법을 시도하다가 결국 addAll 메서드를 만들게 되었는데 지우는 걸 깜빡했네요. 삭제하겠습니다 😄

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -160,6 +176,33 @@ private void removeSections(Section upLineStation, Section downLineStation) {
+    public Sections filteredBy(List<Station> stations) {
+        Sections sections = new Sections();
+        int i = 0;
+        while (i < stations.size()) {
+            if (i != stations.size() - 1) {
```
저는 솔직하게 실무에서는 반복문 내에 if문을 작성하곤 합니다 😂 
아쉽지만 우테켐 과정에서는 depth 1(반복문 내부에 if문 허용 x)만 허용하기에 
해당 부분을 리팩토링 해주시면 좋을 것 같아요 :)

ssamzag replied:
넵 depth1 이 되도록 수정할게요!

ssamzag replied:
제가 짰지만 사실 맘에 들지 않는 코드입니다.. 우선 depth 1로 줄여놓고 다시 생각해 봐야겠네요. 😢

ohtaeg commented on src/main/java/nextstep/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.List;
+
+@Service
+@Transactional(readOnly = true)
+public class PathService {
```
LineService에서는 아래와 같이 되어있는데요!
혹시 서비스마다 트랜잭션을 다르게 설정해주시는 기준이 있으신가요 :) ?
```
@Transactional
public class LineService {
```

ssamzag replied:
서비스의 모든 메서드가 readonly라는 의도로 클래스 readonly를 붙였습니다. 
제가 자바 스프링을 처음 써 봐서 그러는데 혹시 현업에서는 고정적으로 클래스 @transactional 어너테이션을 쓰는 건지 궁금해요!!

ohtaeg replied:
아하 그러시군요 :) 우선 보통 데이터베이스에 연산을 하는 서비스라면 트랜잭션을 관리하기 위해 어노테이션을 붙여줍니다!
예로 데이터베이스로부터 접근하여 어떤 작업을 진행하는 중 로그는 꼭 DB에 저장해야하는데
만약 로그 저장하는 작업이 실패할 경우 메인 작업의 트랜잭션까지 롤백되버린다면 중요한 이슈라고 생각하는데요 :)
혹은 로그를 남긴 후 메인 작업에서 예외가 발생하면 저장한 로그도 롤백되어야 하는 경우
메인 트랜잭션에서 분리해서 중첩 트랜잭션으로 만들어두는 관리 등을 위해서 트랜잭션 어노테이션이 필요하다고 생각해요 😄 

-----

위 리뷰의 의도는 PathService는 readonly가 설정되어있고 LineService에 readonly가 설정이 되어 있지 않아서
혹시 다른 의도가 있으셨는지 궁금해서 여쭤보았었어요 :)

클래스 레벨에서 `@Transactional(readOnly = true)` 으로 설정한다면 더티 체킹을 방지할 수 있어서 
조금이나마 성능 향상에 이점이 있다고 생각하는데요 :) 
더티 체킹이 발생하지 않으므로 엔티티 다루는 과정에서 실수로 발생할 수 있는 변경을 막을 수 있다고 생각도 하구요!

그래서 저는 클래스 레벨에 `@Transactional(readOnly = true)` 을 사용해 기본적으론 읽기 전용 트랜잭션을 사용하고, 
생성, 추가, 삭제와 같은 변경이 필요한 시점에만 메소드 레벨에서  명시적으로 `@Transactional` 을 붙여 줍니다. 😄

ssamzag replied:
말씀하신대로 클래스 레벨에 readOnly 옵션을 붙어주는 게 장점이 많은 거 같네요. 
앞으론 이 기준으로 transaction을 붙여줘야겠습니다. 감사합니다. 😄

ohtaeg commented on src/main/java/nextstep/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,40 @@
+        Station target = stationService.findStationById(targetId);
+        List<Station> stations = stationRepository.findAll();
+        Sections sections = lineService.findAllSections();
+        List<Station> shortestStations = PathFinder.dijkstra(source, target, stations, sections);
```
stations와 sections를 같이 파라미터로 전달해주고 있군요!
저는 Sections를 통해 stations를 알 수 있다고 판단하였는데, 별도로 station을 또 조회해주시는 의도가 궁금한데요 :)
혹시나 의도를 알려주신다면 감사하겠습니다 😄

ssamzag replied:
Sections의 stations 메서드는 라인서비스의 getStations()를 리팩토링 한 것인데 하나의 라인을 기준으로 역들을 가져오도록 만들어진 거 같더라구요. 이 메서드를 고치는 건 이번 미션의 의도를 벗어나는 거 같아 건드리지 않았습니다. 😢

ssamzag replied:
한 줄의 코드로 끝낼 수 있어서 stations를 가져왔어요.
Sections에 구현된 stations()는 라인서비스의 getStations()를 리팩토링한 것인데 라인 1개를 기준으로 만들어진 것 같더라구요. 그래서 빠지는 역들이 엄청 생기네요.
그렇다고 라인별로 지하철역을 가져오도록 개발하자니 코드가 늘어나서 단순하게 db에서 조회하도록 했어요. 물론 HashSet에 상행/하행역을 때려박으면 되긴하겠네요! 😯

ohtaeg replied:
아하 그렇네요! Line과 Sections가 연관 관계가 맺어져 있어서 하나의 라인 기준에 대한 Sections들을 갖고 오겠군요 :)
이해했습니다 👍 👍 이 부분은 제가 꼼꼼하게 못봤네요 🙇

ssamzag replied:
덕분에 HashSet이 떠 올라서 적용시켜 보았습니다. 😃 
서울 지하철 기준으로 DB 쿼리날리는 것과 비교하면 성능비용 차이가 궁금해지네요. 😨

ohtaeg commented on src/main/java/nextstep/subway/path/domain/PathFinder.java:
```diff
@@ -0,0 +1,27 @@
+
+import java.util.List;
+
+public class PathFinder {
```
이 부분은 같이 고민해보면 좋을 것 같아요!
외부 라이브러리를 이용한 최단 경로 찾는 방법은 언제든지 바뀔 수 있는 점을 고려해보았을 때
만약 다른 최단경로알고리즘 라이브러리를 사용하거나 직접 구현할 경우 알고리즘 교체에 유연한지 고민해보면 좋을 것 같아요 :)

인터페이스를 선언하고, 실제 구현체는 외부에서 주입받아 외부 라이브러리 구현체에 직접 의존하는 것을 최소화하면 어떨까요?
민호님은 이 부분에 대해서는 어떻게 생각하시나요 :) ?

ssamzag replied:
OCP 를 준수해서 연습해보란 말씀이시군요!
음 이건 Strategy 패턴을 사용해서 구현하도록 할게요! 😄

ohtaeg commented on src/main/java/nextstep/subway/path/domain/SubwayGraph.java:
```diff
@@ -0,0 +1,28 @@
+
+import java.util.List;
+
+public class SubwayGraph extends WeightedMultigraph<Station, DefaultWeightedEdge> {
```
👍 👍

ohtaeg commented on src/main/java/nextstep/subway/path/ui/PathController.java:
```diff
@@ -0,0 +1,37 @@
+        return ResponseEntity.ok().body(response);
+    }
+
+    @ExceptionHandler(DataIntegrityViolationException.class)
+    public ResponseEntity handleIllegalArgsException(DataIntegrityViolationException e) {
+        return ResponseEntity.badRequest().build();
+    }
```
사소하지만 LineController에서도 예외 처리 핸들러가 중복이 되는 것 같아요..!
중복을 제거하는 것도 좋아보이는데, 민호님은 어떻게 생각하시나요 :) ?
만약 제거한다면, 어떤 방법으로 제거할 수 있을지 고민해보아요 😄

ssamzag replied:
클래스 상속이나 advice에서 처리를 하는 게 좋을 것 같아요. 
찾아보니 @ControllerAdvice 가 있군요. 
자바 스프링.. 편하군요 👍

ohtaeg commented on src/test/java/nextstep/subway/path/PathAcceptanceTest.java:
```diff
@@ -1,9 +1,122 @@
+    private StationResponse 교대역;
+    private StationResponse 남부터미널역;
+
+    /**
+     * 교대역    --- *2호선* ---   강남역   --- *2호선* ---  력삼역
+     * |                        |
+     * *3호선*                   *신분당선*
+     * |                        |
+     * 남부터미널역  --- *3호선* ---   양재
+     */
```
👍 👍

ohtaeg commented on src/test/java/nextstep/subway/path/PathAcceptanceTest.java:
```diff
@@ -1,9 +1,122 @@
+        super.setUp();
+
+        강남역 = StationAcceptanceTest.지하철역_등록되어_있음("강남역").as(StationResponse.class);
+        력삼역 = StationAcceptanceTest.지하철역_등록되어_있음("력삼역").as(StationResponse.class);
```
💪

ssamzag replied:
💪삼💪

ohtaeg commented on src/test/java/nextstep/subway/path/PathAcceptanceTest.java:
```diff
@@ -1,9 +1,122 @@
+        이호선 = LineAcceptanceTest.지하철_노선_등록되어_있음(new LineRequest(
+                "이호선", "bg-red-600", 교대역.getId(), 강남역.getId(), 1000)).as(LineResponse.class);;
+        삼호선 = LineAcceptanceTest.지하철_노선_등록되어_있음(new LineRequest(
+                "삼호선", "bg-red-600", 교대역.getId(), 남부터미널역.getId(), 5)).as(LineResponse.class);;
+
+        지하철_노선에_지하철역_등록_요청(삼호선, 교대역, 남부터미널역, 3);
```
사소하지만 구간 거리가 다르군요 👀

ssamzag replied:
억 중복이군요.. 삭제하겠습니다 😅

ohtaeg commented on src/test/java/nextstep/subway/path/PathAcceptanceTest.java:
```diff
@@ -1,9 +1,122 @@
+        PathResponse pathResponse = response.as(PathResponse.class);
+
+        //then
+        assertThat(pathResponse.getStations()).hasSize(4);
+        assertThat(pathResponse.getStations().get(0).getId()).isEqualTo(교대역.getId());
+        assertThat(pathResponse.getStations().get(1).getId()).isEqualTo(남부터미널역.getId());
+        assertThat(pathResponse.getStations().get(2).getId()).isEqualTo(양재역.getId());
+        assertThat(pathResponse.getStations().get(3).getId()).isEqualTo(강남역.getId());
+        assertThat(pathResponse.getDistance()).isEqualTo(18);
```
사소하지만 여러 검증문을 검증하길 원하신다면 assertAll()을 이용하시는 것도 좋아보여요 :)

ssamzag replied:
넵 적용해 보겠습니다!

ohtaeg commented on src/test/java/nextstep/subway/path/PathAcceptanceTest.java:
```diff
@@ -1,9 +1,122 @@
+import static org.assertj.core.api.Assertions.assertThat;
 
 @DisplayName("지하철 경로 조회")
 public class PathAcceptanceTest extends AcceptanceTest {
```
전체적으로 테스트를 잘 작성해주셨네요 👍 
개인적으로 추가로 더 테스트 해봤는데 다 통과하네요 👍

ohtaeg reviewed (changes requested):
2단계도 구현을 잘해주셨네요 👍 👍 
미션 요구사항을 잘 만족해주셔서 넘어갈까하다가..!
같이 고민해보면 좋을 포인트가 있어서 Request changes 하였습니다 :)
소소한 코멘트를 드렸는데 확인해주시면 감사하겠습니다 🙏 

p.s. 덕분에 생각보다 빨리 칼퇴(?) 했습니다 💤

ssamzag committed:
refactor: 코드리뷰 반영
- 안 쓰는 코드 제거
- Transaction Readonly 레벨을 클래스 단위로 적용
- Stations 를 Sections 객체에서 가져오도록 변경
- @ControllerAdvice 구현
- 코드리뷰 링크
https://github.com/next-step/atdd-subway-service/pull/618

ssamzag committed:
refactor: filteredBy 메서드 코드 depth 1로 수정

ssamzag committed:
docs: 1단계, 2단계 요구사항 및 구현목록 정리

ohtaeg commented on src/main/java/nextstep/subway/ControllerExceptionHandler.java:
```diff
@@ -0,0 +1,20 @@
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
```
`@ControllerAdvice`를 잘 적용해주셨어요 👍 👍 
아시는 내용일 수도 있겠지만 모르실 수도 있기에 `@RestControllerAdvice`라는 것도 있는데요!
두 어드바이스는 어떤 차이도 있는지 따로 정리해보시면 좋을 것 같아요 😄

ssamzag replied:
ResponseBody로 처리하는 게 맞겠네요; 
감사합니다 😄

ohtaeg commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -13,7 +13,7 @@
 
 @Service
-@Transactional
+@Transactional(readOnly = true)
```
👍

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -76,14 +72,25 @@ private void updateDownStationIfPresent(Station upStation, Station downStation,
 
-    public List<Station> stations() {
+    public List<Station> allStations() {
+        HashSet<Station> stations = new HashSet<>();
+
+        for (Section s : sections) {
+            stations.add(s.getUpStation());
+            stations.add(s.getDownStation());
+        }
+
+        return new ArrayList<>(stations);
```
> 덕분에 HashSet이 떠 올라서 적용시켜 보았습니다. 😃

중복되지 않는 역을 담기 위해 Set을 사용하셨군요 👍 👍 
중복되지 않는 값을 찾기 위해서 Set을 이용하는 방법은 매우 좋다고 생각해요 💯 

추가로 중복되지 않는 역을 찾는 다른 방법도 떠올랐는데요 :)
List를 반환하고 계시기에 Set 대신 반환 값인 List를 변수로 추출하여
해당 List에 역을 담고 Stream의 distinct()를 이용하는 방법이 떠올랐어요!

Set을 이용하는 방법보다 좋다고 볼 순 없지만 
중복을 제거하는 여러 방법이 있을 수도 있다는 점을 알려드리고 싶었습니다 :)

수정을 위한 리뷰는 아니오니 참고만 해주시면 감사하겠습니다!

ohtaeg replied:
> 서울 지하철 기준으로 DB 쿼리날리는 것과 비교하면 성능비용 차이가 궁금해지네요. 😨

서울 지하철 기준이라 하시면 서비스 진단하기 미션의 서울교통공사를 말씀 해주시는 걸까요 :) ?
쿼리 실행 계획이 어떻게 될지는 저도 확실히 모르겠지만
어플리케이션 내부에서 성능을 조금이나마 추측해보면, 큰 차이는 얼마 나지 않을 것이라고 생각합니다 ㅎㅎ

HashSet 내부는 Map을 이용하여 값을 보관하기에 시간복잡도가 O(1)이고,
기존 List를 순회하는 방법은 O(N)이라는 차이가 있지만
비지니스상 지하철 역에 대한 데이터는 백만개 데이터 미만의 데이터를 다루기 때문에
어플리케이션상 속도 차이는 크게 안날 것 같기도 하네요 :) 🙇

ssamzag replied:
Stream distance 조합이 코드상으로 깔끔해 보일 것 같아요! 

비용은 건수가 없어서 크게 차이가 나진 않겠군요. 
상세한 답변 감사합니다 😄

ohtaeg commented on src/main/java/nextstep/subway/path/domain/PathFinder.java:
```diff
@@ -0,0 +1,15 @@
+    private PathFinder() {
+    }
+
+    public static List<Station> find(PathStrategy pathStrategy) {
```
피드백 반영 💯

ohtaeg commented on src/main/java/nextstep/subway/path/domain/strategy/Dijkstra.java:
```diff
@@ -0,0 +1,39 @@
+
+import java.util.List;
+
+public class Dijkstra implements PathStrategy {
```
다익스트라 알고리즘을 구현체로 변경 👍

ohtaeg commented on src/main/java/nextstep/subway/path/domain/strategy/PathStrategy.java:
```diff
@@ -0,0 +1,9 @@
+
+import java.util.List;
+
+public interface PathStrategy {
+    List<Station> find();
```
전략 💯

ohtaeg commented on src/main/java/nextstep/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,41 @@
+@Service
+@Transactional(readOnly = true)
+public class PathService {
+    private final StationRepository stationRepository;
```
피드백이 반영됨으로써 해당 Repository는 안쓰게 된 것 같아요 :)

ssamzag replied:
앗 삭제 할게요~

ohtaeg reviewed (approved):
피드백 반영을 잘해주셨어요 👍 👍 
저도 자바를 하다가 이번에는 go를 쓰다보니 조금 낯설기도 한데요!
같이 이번 우테켐을 통해 성장해보아요 :)
소소한 코멘트들을 드렸는데 확인해주시면 감사하겠습니다 🙏 
다음 미션도 화이팅입니다 🔥
